[navigation] Add cycling functionality to SPC TAB and SPC w TAB

fnussbaum opened the pull request:
Sometimes one finds oneself in the situation where one used `SPC TAB` to switch to the last buffer, but it turns out that one actually intended to switch to the second or third to last buffer. By enabling the option added in this commit, one can consecutively press `TAB` in such situations to cycle through the previous buffers. From the point of view of subsequent commands it is as though one switched directly to the destination buffer. This uses the excellent transient-cycles package under the hood (https://elpa.gnu.org/packages/transient-cycles.html).

The setup needs lexical binding in `spacemacs-navigation/packages.el`. I have audited the file and it does not look as if anything depends on dynamic binding.

I considered (and implemented a first version) using hydra/Spacemacs transient states, by the way, but the implementation using the transient-cycles package is cleaner.

The current solution uses somewhat hacky advice to lazily load the package, not sure if this is controversial. Alternatives could be to use remapping instead of advice, or load the package at startup, or perhaps defer it for a few seconds (using `:defer 3`, for example).

fnussbaum committed:
[core] Refactor spacemacs/alternate-buffer

Since Emacs 27, there is the option switch-to-prev-buffer-skip which we can
leverage instead of iterating through the window-prev-buffers ourselves. The
only difference between using previous-buffer and manually switching to the
buffer is that previous-buffer respects and sets window-next-buffers. In order
to keep the same behaviour, we need to set the window's next buffers to nil
before and after invoking previous-buffer. (Note that a window's next buffers are
usually a subset of the window's previous buffers, and switching to a buffer with
any command other than previous-buffer or next-buffer will also set the window's
next buffers to nil.)

This simplifies the function a little bit, and more importantly, enables reusing
the predicate spacemacs//alternate-buffer-skip elsewhere without code
duplication.

Also, deprecate the optional window parameter. It serves no purpose, one can
just as well use with-selected-window at call sites. (And calls using the window
argument will not work with the cycling functionality from the next commit
enabled.)

fnussbaum committed:
[navigation] Add cycling functionality to SPC TAB and SPC w TAB

Sometimes one finds oneself in the situation where one used SPC TAB to switch to
the last buffer, but it turns out that one actually intended to switch to the
second or third to last buffer. By enabling the option added in this commit, one
can consecutively press TAB in such situations to cycle through the previous
buffers. From the point of view of subsequent commands it is as though one
switched directly to the destination buffer. This uses the excellent
transient-cycles package under the
hood (https://elpa.gnu.org/packages/transient-cycles.html).

The setup needs lexical binding in spacemacs-navigation/packages.el. I have
audited the file and it does not look as if anything depends on dynamic binding.

I considered (and implemented a first version) using hydra/Spacemacs transient
states, by the way, but the implementation using the transient-cycles package is
cleaner.

bcc32 committed:
[core] Add warnings for spacemacs/alternate-buffer argument

Add a byte-compilation warning, and also add a runtime warning since
users often don't byte-compile their dotfiles.

bcc32 committed:
[navigation] Remove unnecessary symbol-function

Binding the variables directly to the function name is better for
debuggability, and is terser anyway.

bcc32 commented on core/core-dotspacemacs.el:
```diff
@@ -483,6 +483,35 @@ window."
+Set the option to '(alternate-buffer alternate-window) in order to
+enable cycling for all currently supported commands; or choose a
+subset. (default nil)
+  '(repeat (choice (const alternate-buffer)
```
IMO, it should also be allowed to set this to `t` to enable all current and future cycling commands.  (Will need to adjust the callsites of course to handle when this variable isn't a list)

fnussbaum replied:
Good idea, done.

bcc32 commented on layers/+spacemacs/spacemacs-navigation/packages.el:
```diff
@@ -433,6 +434,88 @@
+       (([remap spacemacs/alternate-window] ()
+         (interactive)
+         (setq sorted-windows
+               (sort (window-list)
```
I think `sort` with keyword arguments is only usable from Emacs 30+ onwards, sadly

fnussbaum replied:
Right, thanks!

bcc32 commented:
> The current solution uses somewhat hacky advice to lazily load the package, not sure if this is controversial. Alternatives could be to use remapping instead of advice, or load the package at startup, or perhaps defer it for a few seconds (using `:defer 3`, for example).

IMO just loading the package at startup is probably fine.  On my box, it takes 4-5ms to load (timed with `(benchmark-run 100 (load "transient-cycles"))`)

BTW I have pushed a couple of small tweaks, I think they should be uncontroversial but if you disagree please lmk.

fnussbaum commented:
> IMO just loading the package at startup is probably fine. On my box, it takes 4-5ms to load (timed with `(benchmark-run 100 (load "transient-cycles"))`)

Makes sense, done.

> BTW I have pushed a couple of small tweaks

Perfect, thank you! I've addressed the feedback, fixed the package's toggle form, and added it to the layer's readme.

fnussbaum committed:
[navigation] Tweak buffer cycling for tab-line-mode

Use push-window-buffer-onto-prev instead of manual consing. This is simpler and
avoids duplicate entries (which do no harm and would get cleaned up lazily, but
can be visible on the tab-line).

Also, immediately update the tab-line after exiting transient cycling.

fnussbaum commented:
I have added another commit to prevent the tab-line from showing an inconsistent state after exiting transient cycling.

smile13241324 reviewed (approved):
This is excellent. Thanks so much for this contribution. The refactoring in core-funcs.el is clean, and the new dotspacemacs-enable-cycling variable is perfectly documented.

I noticed your PR description mentions "somewhat hacky advice," but the implementation I see in the diff (using [remap spacemacs/alternate-buffer]) is the perfectly clean, declarative use-package solution. That's exactly the right way to do it, so I'm assuming the description is just outdated.

Great work! Merging this now.
